### DISCUSSION TITLE
ROT stack op

### DISCUSSION THREAD

**Discussion by pierrec** [maximecb/uvm#35] [Ideas]
Not sure if this is the best name, but I wish the stack implemented a rotation on 3 values on the stack:

```
A
B
C
```
ROT would yields:
```
C
A
B
```

**Comment by maximecb** [comment]
I think some VMs do implement operations of the sort.

Can I ask what is your use case? What are you trying to implement?

**Comment by pierrec** [comment]
I used to work with a stack based calculator and it indeed had the ROT operation.

Running into this from times to times. Lately it is this scenario:
while checking that a slicing operation is correct I end up with this stack:
```
source address # source header
low index
high index
max index
```
to compute the capacity I need to do (max - low), length with (high - low) and save the values to dest and (dest + 8).

Without ROT
```
<get dest>
dup
set_local N
swap
getn 3
sub_u64
store_u64

swap
sub_u64
get_local N
swap
push 8
add_u64
store_u64
```

With ROT
```
getn 2
sub_u64
swap
rot
sub_u64
<get dest>
dup
rot
store_u64
swap
push 8
add_u64
store_u64
```

1 less instruction and no need for a local to store dest. Not a big deal, but I find the second one easier to reason about.

NB. changing the order in which the values are produced would be cumbersome.

**Reply by maximecb** [reply]
I'm a bit hesitant to add `rot` because it's [not very RISC-y](https://www.youtube.com/watch?v=2mzSV3uYbyY).

I did a bit of research and the FORTH language has a `rot` instruction, (and so does the CPython VM, interestingly), but neither do the JVM or the Ruby VM, which kind of goes to show that it's not absolutely necessary to do everything that the JVM and CRuby do.

The downside is that in order to implement `rot`, the VM has to move 3 separate values, and needs a temp register to do so, which seems to encourage generating inefficient code somewhat?

One thing that I could do pretty easily is to add a `setn` instruction to complement `getn`. I know that's not what you asked for, but I do feel like that instruction would be easy to add and no compromise. Would that be helpful?

`setn` would pop the topmost stack value and write the value into a stack slot indexed from the new top. So `setn 0` would be equivalent to removing the element below the current stack top.

**Comment by pierrec** [comment]
Thank you for investigating this. Using a temp register is what I am currently doing. Will see if this can be ridden of by re organising my code.
I think `setn` would definitely help!

**Reply by maximecb** [reply]
Just added `setn` :)

**Comment by pierrec** [comment]
Thanks!!
